controller/src/controller.rs

```diff
@@ -271,27 +271,42 @@ impl<T: BottlerocketShadowClient> BrupopController<T> {
-                    // try to find out if any nodes are being updated. If yes, controller
-                    // will complete them and pause actions on other waitingForUpdate nodes.
-                    if !active_set.is_empty() {
-                        self.progress_active_set(active_set).await?;
-                    } else {
-                        sleep(ACTION_INTERVAL).await;
-                        continue;
-                    }
-                }
-                true => {
-                    if !active_set.is_empty() {
-                        self.progress_active_set(active_set).await?;
-                    }
+            // complete any ongoing updates even if we are outside of a scheduled window.
+            if !active_set.is_empty() {
+                self.progress_active_set(active_set).await?;
+            } else {
+                // calculate the duration between current time and next scheduled time, and then make
+                // controller sleep until next scheduled time point.
+                let next_scheduled_time = brupop_schedule
+                    .upcoming(Utc)
+                    .next()
+                    .context(controllerclient_error::GetScheduledDatetimeSnafu)?;
+                let duration_until_next_schedule = next_scheduled_time - Utc::now();
+
+                // covert chrono duration to std duration
+                let duration_until_next_schedule_std = duration_until_next_schedule
+                    .to_std()
+                    .context(controllerclient_error::ConvertToStdDurationSnafu)?;
+                event!(
+                    Level::INFO,
+                    next_schedule_time = ?next_scheduled_time,
+                    sleep = duration_until_next_schedule_std.as_secs(),
+                    "Began sleeping until next scheduled time point."
+                );
+                tokio::time::sleep(duration_until_next_schedule_std).await;
+            }
+
+            // cron expression is a time-based job scheduler user to run a specified job periodically at fixed times or periods of time (time window).
+            // fixed times: 0 0 10 * * Mon */Every Monday at 10AM.
+            // periods of time (time window): * * 10-12 ? * MON */Every Monday between 10:00 and 12:00.
 
+            // brupop controller needs to use different logics to deal with fixed times or periods of time (time window)
+            // => fixed times: trigger brupop update and complete all waitingForUpdate nodes.
+            // => periods of time (time window): trigger brupop update within time window. If current time isn't within the time window,
+            // controller shouldn't have any action on it.
+            if is_time_window(&brupop_schedule)? {
+                // If current time isn't within the time window, controller shouldn't have any action on it and sleep until next scheduled time point.
+                if within_time_window_cron_expression(&brupop_schedule) {
```

Comment:
I'm confused by these two function calls.

If we've already slept (line 296) for the duration of time up to the next scheduled update, why do we then need to verify the time is within the cron expression? Won't `within_time_window_cron_expression` always be true? Or is there a case where the flow has slept till the next duration and this will return false?

Reply:
A little background  - Main loop has two major parts.
First part is to progress the active nodes on list. 
```
if !active_set.is_empty() {
                self.progress_active_set(active_set).await?;
            } else {
```
Second part is to find active nodes and add to list. 
```
  if active_set_size < get_max_concurrent_update()? {
      // If there's nothing to operate on, check to see if any other nodes are ready for action.
      let new_active_node = self.find_and_update_ready_brs().await?;
      if let Some(brs) = new_active_node {
          event!(Level::INFO, name = %brs.name_any(), "Began updating new node.")
      }
  }
```

Back to your question. For maintenance window (time window), there is a case that the maintenance window has ended, but there are still few active nodes on list. The strategy here is we stop adding nodes to the list of active nodes and complete remaining workflow.  Therefore function `within_time_window` is helping to figure out if controller should add new active nodes to the list when users enable cron expression - maintenance window.